libclingo/src/clingocontrol.cc

```diff
@@ -392,12 +393,46 @@ class ClingoPropagateInit : public PropagateInit {
     void addWatch(uint32_t solverId, Lit_t lit) override { p_.addWatch(solverId, Clasp::decodeLit(lit)); }
     void enableHistory(bool b) override { p_.enableHistory(b); };
     Potassco::Lit_t addLiteral() override {
-        ++literals_;
         return Clasp::encodeLit(Clasp::Literal(facade_().ctx.addVar(Clasp::Var_t::Atom), false));
     }
-    void addClause(Potassco::LitSpan lits) override {
-        for (auto &lit : lits) { clauses_.push_back(lit); }
-        clauses_.push_back(0);
+    bool addClause(Potassco::LitSpan lits) override {
+        auto &ctx = static_cast<Clasp::ClaspFacade*>(c_.claspFacade())->ctx;
+        if (ctx.master()->hasConflict()) { return false; }
+        Clasp::ClauseCreator cc{ctx.master()};
+        cc.start();
+        for (auto &lit : lits) {
+            cc.add(Clasp::decodeLit(lit));
+        }
+        return cc.end(Clasp::ClauseCreator::clause_force_simplify).ok();
+    }
+    bool addWeightConstraint(Potassco::Lit_t lit, Potassco::WeightLitSpan lits, Potassco::Weight_t bound, bool eq) override {
+        auto &ctx = static_cast<Clasp::ClaspFacade*>(c_.claspFacade())->ctx;
+        auto &master = *ctx.master();
+        if (master.hasConflict()) { return false; }
+        Clasp::WeightLitVec claspLits;
+        claspLits.reserve(lits.size);
+        Clasp::Var m = 0;
+        for (auto &x : lits) {
+            auto y = Clasp::decodeLit(x.lit);
+            m = std::max(m, y.var());
+            claspLits.push_back({y, x.weight});
+        }
+        if (m > 0 && !master.validVar(m)) {
+            master.acquireProblemVars();
+        }
+        return Clasp::WeightConstraint::create(*ctx.master(), Clasp::decodeLit(lit), claspLits, bound, eq ? Clasp::WeightConstraint::create_eq_bound : 0).ok();
```

Comment:
This looks ok to me but maybe we should consider moving the "acquireProblemVars()" logic to clasp to make it more consistent with clause creation.